docs/module-caching.md

```diff
@@ -0,0 +1,181 @@
+# Modules evaluation and caching
+
+This proposal aims to enforce some guarantees about how many times modules are evaluated and in which contexts. ECMAScript already guarantees that importing the same module from the same file results in a single evaluation, but Module Blocks increase the complexity because they can be passed around.
+
+## Invariants
+
+1. Importing a module with the same string specifier twice from the same file results in a single evaluation (this is already guaranteed by ecma262):
+   ```js
+   // main.js
+   import { check as c1 } from "./file.js";
+   import { check as c2 } from "./file.js";
+   assert(c1 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+   ```js
+   // main.js
+   import { check as c1 } from "./file.js";
+   const { check: c2 } = await import("./file.js");
+   assert(c1 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+   ```js
+   // main.js
+   const { check: c1 } = await import("./file.js");
+   const { check: c2 } = await import("./file.js");
+   assert(c1 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+2. Importing a module with the same string specifier from two different modules defined in the same file and evaluated the same Realm results in a single evaluation:
+   ```js
+   // main.js
+   import { check as c1 } from "./file.js";
+   const mod = module { export { check } from "./file.js"; };
+   const { check: c2 } = await import(mod);
+   assert(c1 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+   ```js
+   // main.js
+   const mod1 = module { export { check } from "./file.js"; };
+   const mod2 = module { export { check } from "./file.js"; };
+   const { check: c2 } = await import(mod1);
+   const { check: c2 } = await import(mod2);
+   assert(c2 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+3. Importing the same module block twice _from the same Realm_, even if from different modules, results in a single evaluation:
```

Comment:
```suggestion
3. Importing the same module block twice _from the same Realm_, even if from different parent modules, results in a single evaluation:
```

Reply:
Now that we can have syntactically nested modules we should come up with two different words for "parent" that represent a parent in the imports linking graph and a parent in the AST.

Reply:
I’m confused; there shouldn’t be any such concept as “parent”, since a module can be imported many times from many different places, including itself. A module just has 0-N importers, which may be static or dynamic. (ref: node’s CJS module.parent which is a massive misfeature for this reason)

Reply:
Yup you are right, the word I needed is "importer".

Reply:
Sounds good - `importer` or `importing` works for me. Perhaps we can update the former `referrer` terminology here as well.